addons/account/models/ir_actions_report.py

```diff
@@ -1,16 +1,87 @@
 # -*- coding: utf-8 -*-
 
-from odoo import models, api, _
+import io
+
+from PIL import Image
+from PyPDF2 import PdfFileWriter, PdfFileReader
+from reportlab.lib import colors
+from reportlab.lib.units import cm
+from reportlab.lib.utils import ImageReader
+from reportlab.pdfgen import canvas
+
+from odoo import models, _
 from odoo.exceptions import UserError
+from odoo.tools import file_open
+
 
 class IrActionsReport(models.Model):
     _inherit = 'ir.actions.report'
 
+    def _add_banner_with_reference(self, record):
+        """ Modify the current PDF file in order to add a banner, in the upper right corner,
+            that will include the Vendor Bill's 'name'. This is mostly done for auditing purposes.
+        """
+
+        file = io.BytesIO(record.message_main_attachment_id.raw)
+        old_pdf = PdfFileReader(file, strict=False, overwriteWarnings=False)
+        old_pdf.getNumPages()
+        packet = io.BytesIO()
+        can = canvas.Canvas(packet)
+        logo = Image.open(file_open('base/static/img/main_partner-image.png', mode='rb'))  # Odoo logo
+        odoo_color = colors.Color(113 / 255, 75 / 255, 103 / 255, 0.8)
+        thickness = 2 * cm
+
+        for p in range(old_pdf.getNumPages()):
+            page = old_pdf.getPage(p)
+            width = float(abs(page.mediaBox.getWidth()))
+            height = float(abs(page.mediaBox.getHeight()))
+
+            can.translate(width, height)
+            can.rotate(-45)
+
+            # Draw banner
+            path = can.beginPath()
+            path.moveTo(-width, -thickness)
+            path.lineTo(-width, -2 * thickness)
+            path.lineTo(width, -2 * thickness)
+            path.lineTo(width, -thickness)
+            can.setFillColor(odoo_color)
+            can.drawPath(path, fill=1, stroke=False)
+
+            # Insert Vendor Bill reference
+            can.setFontSize(10)
+            can.setFillColor(colors.white)
+            can.drawRightString(0.75 * thickness, -1.45 * thickness, record.name)
+            can.drawImage(ImageReader(logo), 0.25 * thickness, -2.05 * thickness, 40, 40, mask='auto', preserveAspectRatio=True)
+
+            can.showPage()
+
+        can.save()
+
+        # Merge the old pages with the matermark
+        watermark_pdf = PdfFileReader(packet, overwriteWarnings=False)
+        new_pdf = PdfFileWriter()
+        for p in range(old_pdf.getNumPages()):
+            new_page = old_pdf.getPage(p)
+            new_page.mergePage(watermark_pdf.getPage(p))
+            new_pdf.addPage(new_page)
+
+        # Write the new pdf into a new attachment
+        with io.BytesIO() as output:
+            new_pdf.write(output)
+            attachment = record.message_main_attachment_id.copy({'raw': output.getvalue()})
```

Comment:
so we're duplicating the attachment every time it gets retrieved? Wasn't it a goal to avoid storing this result anywhere? That looks rather wasteful.
The PDF metadata will be different every time so it won't just be duplicated in `ir.attachment` but also on disk, no?

Reply:
Indeed, extending `_render_qweb_pdf` would help avoiding that